### ISSUE TITLE
Add code generator support for synthetic views

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#11054]
After support for new types of `<syntheticObjects/>` were added to jOOQ 3.14 (#10588), another type of synthetic object could be synthetic `<views/>`. For example:

```xml
<syntheticObjects>
  <views>

    <!-- A classic, static view -->
    <view>
      <catalog>CATALOG</catalog>
      <schema>SCHEMA</schema>
      <name>VIEW</name>
      <comment>Comment</comment>
      <sql>SELECT 1 AS a</sql>
    </view>

    <!-- A table valued function -->
    <view>
      <catalog>CATALOG</catalog>
      <schema>SCHEMA</schema>
      <name>FUNCTION</name>
      <comment>Comment</comment>
      <sql>SELECT id, title FROM book WHERE author_id = :author_id</sql>
    </view>
  </views>
</syntheticObjects>
```

The `VIEW` would behave like an ordinary view, whereas the `FUNCTION` would produce table valued function behaviour as if they were stored like this:

```sql
CREATE OR REPLACE VIEW catalog.schema.view AS select 1 AS a;

-- HSQLDB dialect
CREATE OR REPLACE FUNCTION catalog.schema.function (author_id INTEGER)
RETURNS TABLE (
  id INTEGER,
  title VARCHAR(50)
)
READS SQL DATA
BEGIN ATOMIC
  RETURN TABLE (SELECT id, title FROM book WHERE author_id = $author_id);
END;
```

When referencing the synthetic table or view, the generated SQL would obviously have to define the view in output SQL.

The building blocks introduced with this feature can help implement other features in the future where static SQL statements can be used to generate DAOs and POJOs, etc, see discussion here: https://github.com/jOOQ/jOOQ/issues/11049

**Comment by billbarni** [comment]
Will this features ever be available on the open source jooq license?

**Comment by lukaseder** [comment]
Most of it, yes, see the tag `E: All Editions`

![image](https://user-images.githubusercontent.com/734593/102792465-d4483200-43a8-11eb-9595-9e74b3e0f4b2.png)

Do you have a use-case for this?

**Comment by billbarni** [comment]
But Synthetic unique keys and Synthetic foreign keys will be reserved for professional+ versions?
I'm sorry to bother you with this, I stared working for Philips EMR Brasil (Philips EMR Brazil, a small part of Philips that develops a hospital management software to several parts of the planet).
I am trying to make a demo using jOOQ to convice the crazy people around the shitty 3 different projects that try to replace one another is not the way to go.

I am NOT capable of developing what I want by myself and Philips (specially in Brazil), is a pretty difficult company to change.

I was trying to build a microservice type project (Spring Boot + shenanigans), and every microservice would the jOOQ objects that only it uses. Only the tables and relations that it uses. If something is missing, you add the tables that are missing... or the relations with Synthetic foreing keys... If the database changed the tables/columns that you use, your project breakes and you know where the problem is... while 100% dettatched from other microservices that use their own data sources.

By doing this, you create a solid layer of SQL over a database, using jOOQ that then can be used to migrate a better version (including the Synthetic stuff you added), making a layer that would allow for database migration and control very appreciable.

Other thing that can be done with this Synthetic layer is use GraphQL to build a micro service at the spot using jOOQ generated Synthetic artifacts.

Add some layer somewhere on top/side of this... (null? not null? nullable? allow read? allow write? allow list? owner? - Google Firebase style) and somethign wonderful can be achieved.

I already said it somethimes... DataGeekery could develop something like a GraphQL layer for jOOQ managed objects and synthetic relations built in jOOQ framework. jOOGraph? jOOG?

Well, I'm proably trying to simplify and solve too many problems by just commenting. Thanks a lot for always helping your and you community here and in StackOverflow.

**Comment by lukaseder** [comment]
> But Synthetic unique keys and Synthetic foreign keys will be reserved for professional+ versions?

Yes, but those features are not really related to this one, apart from being "synthetic".

I'm not sure if this stuff is really the best way to do what you want to achieve. Why not work with the `DDLDatabase` instead, and supply "augmented" SQL scripts that already have the "missing" objects? In the near future, we'll be adding more migration tooling to jOOQ, so these things could certainly be made more convenient. For example, you could maintain a branch of your SQL migration that already has the missing objects (and keep merging production changes to that branch), and generate code from the branch, not from main.

I won't re-iterate my comments on GraphQL from elsewhere on this issue here, it's not really related to synthetic views.

**Comment by billbarni** [comment]
Thank you very much for your time and all the clarification. Sorry to missinterpret the relation between because the words used.
Maybe you should use a better naming strategy? (I am just joking, was looking at a youtube video of yours about this, haha).

Thanks you as always!
